# Import and setup

In [1]:
import pandas as pd

In [2]:
# change display options
pd.options.display.max_rows = 150
pd.options.display.max_columns = 150

In [3]:
%%time
# import cleaned tourism and population data
tourism_pop_df = pd.read_csv('../data/derived/tourism_pop.csv', index_col=0)
tourism_pop_df.head(3)

CPU times: user 7.62 ms, sys: 2.08 ms, total: 9.7 ms
Wall time: 9.67 ms


year_wb  country_id country_txt_wb country_code  arrivals  population
0     2012           8         Angola          AGO    528000    25107931
1     2012           5        Albania          ALB   3156000     2900401
2     2012           7        Andorra          AND   2238000       82427

In [4]:
tourism_pop_df.shape

(1078, 6)

The `tourism_pop_df` has 1078 records, representing 7 years of data for 154 countries.

In [5]:
%%time
# import cleaned terrorism data (may take a few minutes to load)
terrorism_df = pd.read_csv('../data/derived/terrorism.csv', index_col=0)
terrorism_df.head(3)

CPU times: user 1.17 s, sys: 177 ms, total: 1.35 s
Wall time: 1.39 s


<string>:2: DtypeWarning: Columns (62,64,80,115,116) have mixed types.Specify dtype option on import or set low_memory=False.


eventid  year_gtd  month_gtd  day_gtd approxdate  extended resolution  \
0  201201010001      2012          1        1        NaN         0        NaN   
1  201201010002      2012          1        1        NaN         0        NaN   
2  201201010004      2012          1        1        NaN         0        NaN   

   country_id country_txt_gtd  region          region_txt      provstate  \
0         153        Pakistan       6          South Asia    Balochistan   
1         104           Kenya      11  Sub-Saharan Africa  North Eastern   
2         104           Kenya      11  Sub-Saharan Africa  North Eastern   

      city  latitude  longitude  specificity  vicinity  \
0   Quetta  30.20082  66.994352          1.0         0   
1  Garissa  -0.45612  39.646123          1.0         0   
2  Garissa  -0.45612  39.646123          1.0         0   

                           location  \
0  Incident occurred on Joint road.   
1                               NaN   
2                               NaN   

                                             summary  crit1  crit2  crit3  \
0  01/01/2012: A grenade was thrown into a shop o...      1      1      1   
1  01/01/2012: Assailants threw two grenades and ...      1      1      1   
2  01/01/2012: Assailants threw a grenade and ope...      1      1      1   

   doubtterr  alternative  alternative_txt  multiple  success  suicide  \
0          0          NaN              NaN         0        1        0   
1          0          NaN              NaN         1        1        0   
2          0          NaN              NaN         1        1        0   

   attacktype1    attacktype1_txt  attacktype2 attacktype2_txt  attacktype3  \
0            3  Bombing/Explosion          NaN             NaN          NaN   
1            2      Armed Assault          NaN             NaN          NaN   
2            2      Armed Assault          NaN             NaN          NaN   

  attacktype3_txt  targtype1 targtype1_txt  targsubtype1     targsubtype1_txt  \
0             NaN          1      Business           2.0  Restaurant/Bar/Café   
1             NaN          1      Business           2.0  Restaurant/Bar/Café   
2             NaN          1      Business           2.0  Restaurant/Bar/Café   

       corp1    target1  natlty1 natlty1_txt  targtype2 targtype2_txt  \
0    Unknown  Paan Shop    153.0    Pakistan        NaN           NaN   
1  Locus Pub        Bar    104.0       Kenya        NaN           NaN   
2    Unknown        Bar    104.0       Kenya        NaN           NaN   

   targsubtype2 targsubtype2_txt corp2 target2  natlty2 natlty2_txt  \
0           NaN              NaN   NaN     NaN      NaN         NaN   
1           NaN              NaN   NaN     NaN      NaN         NaN   
2           NaN              NaN   NaN     NaN      NaN         NaN   

   targtype3 targtype3_txt  targsubtype3 targsubtype3_txt corp3 target3  \
0        NaN           NaN           NaN              NaN   NaN     NaN   
1        NaN           NaN           NaN              NaN   NaN     NaN   
2        NaN           NaN           NaN              NaN   NaN     NaN   

   natlty3 natlty3_txt       gname gsubname gname2 gsubname2 gname3 gsubname3  \
0      NaN         NaN     Unknown      NaN    NaN       NaN    NaN       NaN   
1      NaN         NaN  Al-Shabaab      NaN    NaN       NaN    NaN       NaN   
2      NaN         NaN  Al-Shabaab      NaN    NaN       NaN    NaN       NaN   

                                              motive  guncertain1  \
0                                                NaN          0.0   
1  The specific motive is unknown; however, Al-Sh...          1.0   
2  The specific motive is unknown; however, Al-Sh...          1.0   

   guncertain2  guncertain3  individual  nperps  nperpcap  claimed  claimmode  \
0          NaN          NaN           0   -99.0       0.0      0.0        NaN   
1          NaN          NaN           0   -99.0       0.0      0.0        NaN   
2          NaN  

The previous line threw an error about mixed types on some columns we don't plan to use:
* 62: `gsubname2` | 64: `gsubname3` | 80: `claimmode3_txt` | 115: `divert` | 116: `kidhijcountry`

In [6]:
terrorism_df.shape

(59395, 138)

The `terrorism_df` has 59395 records, representing 59395 terrorist events.

### Anticipate joined dataset sizes

A quick check comparing the number of countries in both datasets...

In [7]:
# find country ids in tourism population dataset
wb_country_ids = set(tourism_pop_df.country_id)
print(f'There are {len(wb_country_ids)} countries in the tourism and population dataset.')

# find country ids in terrorism dataset
gtd_country_ids = set(terrorism_df.country_id)
print(f'There are {len(gtd_country_ids)} countries in the terrorism dataset.')

# find country ids shared by both datasets
shared_country_ids = wb_country_ids.intersection(gtd_country_ids)
print(f'There are {len(shared_country_ids)} countries in both datasets.')

# find country ids unique to tourism population dataset
wb_only_country_ids = wb_country_ids.difference(gtd_country_ids)
print(f'There are {len(wb_only_country_ids)} unique to the tourism and population dataset.')

# find country ids unique to tourism population dataset
gtd_only_country_ids = gtd_country_ids.difference(wb_country_ids)
print(f'There are {len(gtd_only_country_ids)} unique to the terrorism dataset.')

There are 154 countries in the tourism and population dataset.
There are 142 countries in the terrorism dataset.
There are 111 countries in both datasets.
There are 43 unique to the tourism and population dataset.
There are 31 unique to the terrorism dataset.


Since I will be doing a left join of the terrorism data onto the tourism and population data, I expect more than 1078 records (from `tourism_pop_df`) and fewer than 59395 records (from `terrorism_df` in the new dataset. But how many fewer?

Well, there are three join cases to consider here.
1. A country/year has no terrorism incidents. I expect this record from `tourism_pop_df` to be preserved, and NaNs added for fields from `terrorism_df`.
2. A country/year has one terrorism incident. I expect this record from `tourism_pop_df` to be preserved, and non-null data added for fields from `terrorism_df`.
3. A country year has multiple terrorism incidents. I expect this record from `tourism_pop_df` to be duplicated, and non-null data added for fields from `terrorism_df` for every matching record.

We know that there are 43 countries unique to the tourism and population dataset, for 7 years of data, which means that 43x7=301 records fit case 1.

Now I need to quantify cases 2 and 3, to see how many incidents will be preserved, and how many will be duplicated. In order to do that, I need to count the number of records per country and year. For simplicity's sake, I will look at the number of records per country, and accept that the estimate may be off, because I didn't look at year too.

In [8]:
# count incidents per country_id in terrorism dataset
country_id_df = pd.DataFrame(terrorism_df.country_id.value_counts())
country_id_df = country_id_df.reset_index()
country_id_df.rename(columns={'index':'country_id', 'country_id':'terrorism_record_ct'},
                     inplace=True)

# add column for whether country is in tourism_pop_df
country_id_df['in_wb'] = [c in list(wb_country_ids) for c in country_id_df.country_id]

# examine result
country_id_df

country_id  terrorism_record_ct  in_wb
0            95                13557  False
1             4                 8139  False
2           153                 6417  False
3            92                 4043   True
4           147                 3301  False
5           160                 2354   True
6           182                 1910  False
7           228                 1843  False
8           113                 1639  False
9           200                 1599  False
10          205                 1542   True
11           60                 1229   True
12          209                  916   True
13           19                  803  False
14           45                  800   True
15          214                  764   True
16          195                  649   True
17          229                  609  False
18          603                  438   True
19          123                  436   True
20           37                  435  False
21          104                  435  False
22          155                  366   True
23          167                  315   True
24          141                  315   True
25           97                  295   True
26          110                  265   True
27           41                  244  False
28          138                  228   True
29          173                  215   True
30          217                  214   True
31           78                  193   True
32           69                  189   True
33           34                  186  False
34          137                  170   True
35         1004                  166  False
36           93                  156   True
37           75                  149   True
38           18                  135   True
39          183                  110   True
40           33                   94   True
41           43                   89   True
42          146                   73   True
43          186                   72   True
44           44                   72   True
45            6                   69   True
46          198                   66   True
47          158                   65   True
48          130                   58   True
49           65                   56   True
50          208                   47   True
51           98                   44   True
52           94                   43   True
53          121                   42   True
54          203                   37   True
55           14                   34   True
56          213                   33  False
57          159                   33   True
58           42                   33  False
59           38                   30   True
60          222                   28  False
61           96                   27   True
62         1003                   26  False
63          102                   20   True
64          143                   19   True
65          168                   18  False
66           99                   17   True
67           47                   16   True
68           30                   15   True
69          185                   15   True
70          122                   15   True
71          202                   15   True
72           53                   15   True
73           68                   14   True
74          142                   12   True
75           11                   12   True
76           59                   12   True
77           54                   12   True
78          174                   11  False
79           21                   10   True
80           28                   10   True
81          101                   10   True
82          118                    9   True
83          145                    9   True
84          107                    9   True
85           74                    7   True
86          231                    7   True
87          103                    7   True
88           83                    7   True
89           32                    7   True
9

In [9]:
# quick check that the sum of terrorism_record_ct matches the number of records in terrorism_df
sum(country_id_df.terrorism_record_ct) == len(terrorism_df)

True

In [10]:
# how many incidents match case 2?
sum(country_id_df.terrorism_record_ct[country_id_df.in_wb & (country_id_df.terrorism_record_ct==1)])

14

In [11]:
# how many incidents match case 3?
sum(country_id_df.terrorism_record_ct[country_id_df.in_wb & (country_id_df.terrorism_record_ct>1)])

17928

This means that there ought to be approximately 301+14+17928=18243 records in the joined dataframe. We'll keep an eye out for that in a minute.

# Set indices

Before joining the datasets, I will set the indices to the columns that they'll be joined on: `country_id` and `year`.

In [12]:
# rename year column
terrorism_df.rename(columns={'year_gtd':'year'}, inplace=True)
# set columns for join as indices
terrorism_df.set_index(['year', 'country_id'], inplace=True)
# sort by index
terrorism_df.sort_index(inplace=True)
# examine result
terrorism_df.head(3)

eventid  month_gtd  day_gtd approxdate  extended  \
year country_id                                                          
2012 4           201201020007          1        2        NaN         0   
     4           201201020010          1        2        NaN         0   
     4           201201020012          1        1        NaN         0   

                resolution country_txt_gtd  region  region_txt  provstate  \
year country_id                                                             
2012 4                 NaN     Afghanistan       6  South Asia      Logar   
     4                 NaN     Afghanistan       6  South Asia  Nangarhar   
     4                 NaN     Afghanistan       6  South Asia    Badghis   

                         city   latitude  longitude  specificity  vicinity  \
year country_id                                                              
2012 4              Puli Alam  34.002047  69.031149          1.0         1   
     4             Paindakhel  34.343559  67.921173          2.0         0   
     4           Bala Morghab  35.585669  63.328089          1.0         0   

                               location  \
year country_id                           
2012 4           Khwaja Azfal Qala area   
     4                              NaN   
     4                              NaN   

                                                           summary  crit1  \
year country_id                                                             
2012 4           01/02/2012: A civilian car hit a roadside bomb...      1   
     4           01/02/2012: A suicide bomber detonated near a ...      1   
     4           01/01/2012: Two assailants died when their min...      1   

                 crit2  crit3  doubtterr  alternative  alternative_txt  \
year country_id                                                          
2012 4               1      1          0          NaN              NaN   
     4               1      1          0          NaN              NaN   
     4               1      1          0          NaN              NaN   

                 multiple  success  suicide  attacktype1    attacktype1_txt  \
year country_id                                                               
2012 4                  0        1        0            3  Bombing/Explosion   
     4                  0        1        1            3  Bombing/Explosion   
     4                  0        1        0            3  Bombing/Explosion   

                 attacktype2 attacktype2_txt  attacktype3 attacktype3_txt  \
year country_id                                                             
2012 4                   NaN             NaN          NaN             NaN   
     4                   NaN             NaN          NaN             NaN   
     4                   NaN             NaN          NaN             NaN   

                 targtype1                targtype1_txt  targsubtype1  \
year country_id                                                         
2012 4                  14  Private Citizens & Property          67.0   
     4                   3                       Police          22.0   
     4                  14  Private Citizens & Property          67.0   

                                                targsubtype1_txt  \
year country_id                                                    
2012 4                              Unnamed Civilian/Unspecified   
     4           Police Building (headquarters, station, school)   
     4                              Unnamed Civilian/Unspecified   

                          corp1            target1  natlty1  natlty1_txt  \
year country_id                                                            
2012 4           Not Applicable          Civilians      4.0  Afghanistan   
     4            Afghan Police  Training Facility      4.0  Afghanistan   
     4           Not Applicable          Civilians      4.0  Afghanistan   

                 targtype2     

In [13]:
# rename year column
tourism_pop_df.rename(columns={'year_wb':'year'}, inplace=True)
# set columns for join as indices
tourism_pop_df.set_index(['year', 'country_id'], inplace=True)
# sort by index
tourism_pop_df.sort_index(inplace=True)
# examine result
tourism_pop_df.head(3)

country_txt_wb country_code  arrivals  population
year country_id                                                  
2012 5                 Albania          ALB   3156000     2900401
     6                 Algeria          DZA   2634000    37383887
     7                 Andorra          AND   2238000       82427

# Join datasets

I'll be doing a left join between these two datasets, to only keep countries and years that have tourism/population data. If a country and year is listed in `tourism_pop_df` and not `terrorism_df`, the columns related to terrorism will have NaN values.

In [14]:
# join tourism, population and terrorism data
df = tourism_pop_df.join(terrorism_df, how='left')
# reset index
df.reset_index(inplace=True)
# examine result
df.head()

year  country_id country_txt_wb country_code  arrivals  population  \
0  2012           5        Albania          ALB   3156000     2900401   
1  2012           6        Algeria          DZA   2634000    37383887   
2  2012           6        Algeria          DZA   2634000    37383887   
3  2012           6        Algeria          DZA   2634000    37383887   
4  2012           6        Algeria          DZA   2634000    37383887   

        eventid  month_gtd  day_gtd approxdate  extended  resolution  \
0           NaN        NaN      NaN        NaN       NaN         NaN   
1  2.012012e+11        1.0     16.0        NaN       1.0  2012-01-17   
2  2.012022e+11        2.0     19.0        NaN       0.0         NaN   
3  2.012030e+11        3.0      3.0        NaN       0.0         NaN   
4  2.012033e+11        3.0     26.0        NaN       0.0         NaN   

  country_txt_gtd  region                  region_txt     provstate  \
0             NaN     NaN                         NaN           NaN   
1         Algeria    10.0  Middle East & North Africa        Biskra   
2         Algeria    10.0  Middle East & North Africa     Boumerdes   
3         Algeria    10.0  Middle East & North Africa  Tamanghasset   
4         Algeria    10.0  Middle East & North Africa    Tizi Ouzou   

           city   latitude  longitude  specificity  vicinity  \
0           NaN        NaN        NaN          NaN       NaN   
1        Debdeb  34.840833   6.041389          1.0       1.0   
2       Unknown  36.759396   3.470261          4.0       0.0   
3  Tamanghasset  22.784648   5.522710          1.0       0.0   
4   Tizi Ghenif  36.586090   3.767982          1.0       0.0   

                                  location  \
0                                      NaN   
1                                      NaN   
2  Road 12 between Isser and Bordj Menaiel   
3                                      NaN   
4                                      NaN   

                                             summary  crit1  crit2  crit3  \
0                                                NaN    NaN    NaN    NaN   
1  01/16/2012: Three gunmen abducted Mohamed el A...    1.0    1.0    1.0   
2  02/19/2012: A roadside bomb detonated, targeti...    1.0    1.0    1.0   
3  03/03/2012: A suicide bomber detonated an expl...    1.0    1.0    1.0   
4  03/26/2012: An explosive device detonated on a...    1.0    1.0    1.0   

   doubtterr  alternative  alternative_txt  multiple  success  suicide  \
0        NaN          NaN              NaN       NaN      NaN      NaN   
1        0.0          NaN              NaN       0.0      1.0      0.0   
2        0.0          NaN              NaN       0.0      1.0      0.0   
3        0.0          NaN              NaN       0.0      1.0      1.0   
4        0.0          NaN              NaN       0.0      1.0      0.0   

   attacktype1              attacktype1_txt  attacktype2 attacktype2_txt  \
0          NaN                          NaN          NaN             NaN   
1          6.0  Hostage Taking (Kidnapping)          NaN             NaN   
2          3.0            Bombing/Explosion          NaN             NaN   
3          3.0            Bombing/Explosion          NaN             NaN   
4          3.0            Bombing/Explosion          NaN             NaN   

   attacktype3 attacktype3_txt  targtype1         targtype1_txt  targsubtype1  \
0          NaN             NaN        NaN                   NaN           NaN   
1          NaN             NaN        2.0  Government (General)          15.0   
2          NaN             NaN        4.0              Military          29.0   
3          NaN             NaN        3.0                Police          22.0   
4          NaN             NaN        3.0                Police          25.0   

                                    targsubtype1_txt  \
0                                                NaN   
1  Politician or Political Party Movement/Meeting...   
2                     

In [15]:
# write to csv
df.to_csv('../data/derived/combined_data.csv')

In [16]:
df.shape

(18540, 142)

The dataset now has 18540 records, representing 18540 terrorist incidents that occurred in countries that are included in the tourism and population dataset. This is close to my earlier approximation of 18243 records.

# Group by incident

If an event occurs at separate times or locations, than they are considered separate incidents; see p. 12 of codebook. "If any discontinuity in time or space is noted, the event is comprised of multiple incidents."

If the event is part of a multiple incident, the variable `related` includes the text of event ID's from `eventid` for related incidents; see p. 13 and 17 of codebook.

I will group events into one incident if records are related, occurred in the same city, and occured on the same day. Numbers of fatalities, injured, and victims are summed across records.

In [17]:
df.related.count()

3340

There are 3340 incidents that are related, so I expect to lose around 3340 records by collapsing records. Let's keep that in mind for after the group by and aggregation.

In [18]:
# fill `related` column with `eventid` for singular events (needed for group by)
index_to_fill = df.index[df.eventid.notna() & df.related.isna()]
df.iloc[index_to_fill, 138] = df.iloc[index_to_fill, 6]

# check that `related` column is filled
df.related.count()

17942

In [19]:
df.shape

(18540, 142)

In [20]:
# group related incidents that occur on same day in same city
df_by_incident = df.groupby(['related',
                             'year','month_gtd','day_gtd',
                             'country_id','provstate','city'], dropna=False).agg(
    
    country_txt_wb  = ('country_txt_wb',  'first'),
    country_txt_gtd = ('country_txt_gtd', 'first'),
    country_code    = ('country_code',    'first'),
    region          = ('region',          'first'),
    region_txt      = ('region_txt',      'first'),
    arrivals        = ('arrivals',        'first'),
    population      = ('population',      'first'),
    incident_id     = ('eventid',         'first'),
    
    nkill     = ('nkill',     'sum'),
    nkillus   = ('nkillus',   'sum'),
    nkillte   = ('nkillte',   'sum'),
    nwound    = ('nwound',    'sum'),
    nwoundus  = ('nwoundus',  'sum'),
    nwoundte  = ('nwoundte',  'sum'),
    nvictim   = ('nvictim',   'sum'),
    nvictimus = ('nvictimus', 'sum'),
    nvictimte = ('nvictimte', 'sum'),
    
    event_ct_total   = ('eventid', 'count')
)

In [21]:
# reset index
df_by_incident.reset_index(inplace=True)

In [22]:
# drop columns
df_by_incident.drop(columns=['related','month_gtd','day_gtd','provstate','city'], inplace=True)

In [23]:
# inspect dataframe
df_by_incident.head()

year  country_id country_txt_wb country_txt_gtd country_code  region  \
0  2012          92          India           India          IND     6.0   
1  2012         217  United States   United States          USA     1.0   
2  2012          92          India           India          IND     6.0   
3  2012         141          Nepal           Nepal          NPL     6.0   
4  2012         214        Ukraine         Ukraine          UKR     9.0   

       region_txt  arrivals  population   incident_id  nkill  nkillus  \
0      South Asia   6578000  1265782790  2.012010e+11    0.0      0.0   
1   North America  66966900   313830990  2.012010e+11    0.0      0.0   
2      South Asia   6578000  1265782790  2.012010e+11    0.0      0.0   
3      South Asia    803000    26989862  2.012010e+11    0.0      0.0   
4  Eastern Europe  23013000    45593342  2.012010e+11    1.0      0.0   

   nkillte  nwound  nwoundus  nwoundte  nvictim  nvictimus  nvictimte  \
0      0.0     1.0       0.0       0.0      1.0        0.0        0.0   
1      0.0     0.0       0.0       0.0      0.0        0.0        0.0   
2      0.0     3.0       0.0       0.0      3.0        0.0        0.0   
3      0.0     1.0       0.0       0.0      1.0        0.0        0.0   
4      0.0     1.0       0.0       0.0      2.0        0.0        0.0   

   event_ct_total  
0               1  
1               1  
2               1  
3               1  
4               1

In [24]:
df_by_incident.shape

(17252, 20)

I see that collapsing records into related incidents has led to a loss of 18540-17252=1288 records. This seems reasonable, because not all related events occurred in the same city on the same day, which is the criteria used here.

In [25]:
# write to csv
df_by_incident.to_csv('../data/derived/combined_data_by_incident.csv')

# Group by year

Now, I will collapse by country and year. This should lead to a dataframe with the same number of records as the `tourism_pop_df`: 1078.

In [26]:
# group by year and country
df_by_year = df_by_incident.groupby(['year','country_id']).agg(
    
    country_txt_wb  = ('country_txt_wb',  'first'),
    country_txt_gtd = ('country_txt_gtd', 'first'),
    country_code    = ('country_code',    'first'),
    region          = ('region',          'first'),
    region_txt      = ('region_txt',      'first'),
    arrivals        = ('arrivals',        'first'),
    population      = ('population',      'first'),
    
    nkill_total       = ('nkill', 'sum'),
    nkillus_total     = ('nkillus', 'sum'),
    nkillte_total     = ('nkillte', 'sum'),
    nwound_total      = ('nwound', 'sum'),
    nwoundus_total    = ('nwoundus', 'sum'),
    nwoundte_total    = ('nwoundte', 'sum'),
    nvictim_total     = ('nvictim', 'sum'),
    nvictimus_total   = ('nvictimus', 'sum'),
    nvictimte_total   = ('nvictimte', 'sum'),
    incident_ct_total = ('incident_id', 'count')
)

# reset index
df_by_year.reset_index(inplace=True)

# inspect dataframe
df_by_year.head()

year  country_id       country_txt_wb country_txt_gtd country_code  region  \
0  2012           5              Albania            None          ALB     NaN   
1  2012           6              Algeria         Algeria          DZA    10.0   
2  2012           7              Andorra            None          AND     NaN   
3  2012           8               Angola            None          AGO     NaN   
4  2012          10  Antigua and Barbuda            None          ATG     NaN   

                   region_txt  arrivals  population  nkill_total  \
0                        None   3156000     2900401          0.0   
1  Middle East & North Africa   2634000    37383887         25.0   
2                        None   2238000       82427          0.0   
3                        None    528000    25107931          0.0   
4                        None    247000       90409          0.0   

   nkillus_total  nkillte_total  nwound_total  nwoundus_total  nwoundte_total  \
0            0.0            0.0           0.0             0.0             0.0   
1            0.0            2.0          76.0             0.0             0.0   
2            0.0            0.0           0.0             0.0             0.0   
3            0.0            0.0           0.0             0.0             0.0   
4            0.0            0.0           0.0             0.0             0.0   

   nvictim_total  nvictimus_total  nvictimte_total  incident_ct_total  
0            0.0              0.0              0.0                  0  
1          101.0              0.0              2.0                 27  
2            0.0              0.0              0.0                  0  
3            0.0              0.0              0.0                  0  
4            0.0              0.0              0.0                  0

In [27]:
df_by_year.shape

(1078, 19)

There are once again 1078 records, as expected.

In [28]:
# write to csv
df_by_year.to_csv('../data/derived/combined_data_by_year.csv')